[#35],[#36] Study Room 퇴장 Service 생성 , Study Room 입장/퇴장 API

wooni97 opened the pull request:
- Study Room 퇴장 비즈니스 로직과 관련한 StudyRoomExitService를 추가합니다.
- 웹소캣을 이용한 실시간 입장, 퇴장 인원 확인 기능을 하는 API를 추가합니다. 
- Stomp 엔드포인트를 통해 아래와 같은 기능을 하도록 구현했습니다.
  - 입장 및 퇴장 시 메시지 전송 
  - 방장 권한 위임 기능
  - 방 정보 업데이트하여 클라이언트로 전송
  - 참여자 목록 업데이트하여 클라이언트로 전송

#35 #36

wooni97 committed:
chore: build.gradle에 WebSocket 의존성 추가

wooni97 committed:
feat: StompWebSocketConfig 클래스 추가

wooni97 committed:
chore: build.gradle에 implementation 추가

wooni97 committed:
feat: Cookie 추가하도록 login 메서드 변경

- 로그인 응답에 'memberID'와 'sequecneID' 쿠키 추가
- 프론트 엔드에서 사용하기 위한 조치

wooni97 committed:
refactor: Participant 도메인 클래스 수정

- entryTime 필드 추가
- 방장 권한 위임시 방장 다음으로 가장 빠르게 입장한 유저를 선택하기 위한 필드 추가

wooni97 committed:
refactor: 웹소캣 핸드셰이크 경로 수정

wooni97 committed:
feat: StudyRoom 클래스 메서드 추가

- 방장 변경 메서드 추가
- 방장인지 확인하는 메서드 추가

wooni97 committed:
feat: StompStudyRoomController 클래스 생성

- /room/enter, /room/exit, /room/host-authorization/delegate, /room/info, /room/participants 엔드포인트 추가
- 입장 및 퇴장 시 메시지 전송 및 참여자 목록 업데이트
- 방장 권한 위임 기능 추가
- 스터디룸 정보 및 참여자 목록 클라이언트로 전송하는 메서드 추가

wooni97 committed:
feat: StudyRoomMessage DTO 클래스 생성

- 스터디룸 ID, 유저 ID, 유저 시퀀스 ID, 메세지를 담는 DTO 클래스 추가

wooni97 committed:
feat: 웹 소캣 및 StudyRoom 관련 기능 추가

- 웹 소켓을 사용하여 실시간으로 방 정보 및 참가자 업데이트 기능 추가
- 방 입장 및 퇴장 메시지 표시 기능 추가
- 방장 권한 위임 및 버튼 디스플레이 관련 기능 추가
- 참가자 목록 동적 업데이트 기능 추가
- 클라이언트 쿠키를 활용한 멤버 식별 기능 추가
- 사용자가 방을 나갈 때 DELETE 요청을 보내고 웹 소켓 메시지 전송
- 페이지 로드 및 언로드 이벤트에 대한 처리 추가 추웹

wooni97 committed:
feat: AuthDelegateDTO 클래스 생성

- 방장 권한 위임을 위한 정보를 담은 DTO 클래스 추가

wooni97 committed:
feat: ParticipantRepository 메서드 추가

wooni97 committed:
feat: ParticipantResponse 클래스 생성

- Particiapnt 의 정보를 담은 Response DTO 클래스 추가

wooni97 committed:
feat: StudyRoom API 추가

- roomID로 받은 방의 정보를 가져오는 API 추가

wooni97 committed:
refactor: 데이터베이스 스키마 변경

- STUDY_ROOM 테이블에 max_participants 및 current_participants 열 추가
- PARTICIPANT 테이블에 entry_time 열 추가
- 스키마 변경에 따른 repository 변경

wooni97 committed:
feat: StudyRoomExitService 클래스 생성

- Study Room 퇴장 관련 비즈니스 로직을 모은 서비스 클래스 생성
- 방장 권한을 가진 유저일 경우 권한을 위임 할 유저를 찾아 권한 위임 후 퇴장

f-lab-inno commented on src/main/resources/templates/rooms.html:
```diff
@@ -0,0 +1,58 @@
+        // 데이터 구조에 따라 처리
+        if (Array.isArray(apiResponse.data)) {
+            apiResponse.data.forEach(room => {
+                const roomItem = document.createElement('div');
+                roomItem.innerHTML = `
+                <p>Room ID: ${room.roomId}</p>
+                <p>Room Name: ${room.roomName}</p>
+                <p>Max Participants: ${room.maxParticipants}</p>
+                <p>Current Participants: ${room.currentParticipants}</p>
+                <p>Room Manager Sequence ID: ${room.roomManagerSequenceId}</p>
+                <button onclick="joinRoom(${room.roomId})">Join Room</button>
+                <hr>
+            `;
```
이왕 jquery를 추가하셨으니 dom 핸들링 부분은 jquery로 해보는것도 좋을것 같네요! (반영하실필요는 없습니다)
바닐라JS로 구현하신 부분도 인상적이에요~!👍

f-lab-inno commented on src/main/resources/templates/room.html:
```diff
@@ -0,0 +1,221 @@
+      roomManagerSequenceId = data.data.roomManagerSequenceId;
+    } catch (error) {
+      console.error('Error during fetch:', error);
+    }
```
ES6 표준 문법을 사용해주셨네요 👍

f-lab-inno commented on src/main/resources/templates/room.html:
```diff
@@ -0,0 +1,221 @@
+  <button id="exitButton">나가기</button>
+</div>
+<script>
+  var roomManagerSequenceId;
```
특별한 이유가 없다면 요것처럼 var 선언보다는 ES6 표준 변수 선언이 사용되어도 좋을것 같아요.

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/room/service/StudyRoomService.java:
```diff
@@ -59,6 +60,14 @@ public StudyRoom enterRoom(int roomId, int memberSequenceId) {
     public StudyRoom exitRoom(int roomId, int memberSeqId){
         StudyRoom studyRoom = studyRoomRepository.findByRoomId(roomId).orElseThrow();
         participantRepository.delete(studyRoom.getRoomId(), memberSeqId);
+
+        studyRoom.exitRoom();
+        studyRoomRepository.update(studyRoom.getRoomId(),
+                studyRoom.getRoomName(),
+                studyRoom.getMaxParticipants(),
+                studyRoom.getCurrentParticipants(),
+                studyRoom.getManagerSequenceId());
+
```
`StudyRommExitService.exitRoom` 과는 다른 기능일까요?

wooni97 replied:
아뇨! 퇴장 관련 로직들을 StudyRoomExitService로 따로 분리를 하면서 지웠어야 하는 코드였습니다.

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/room/service/StudyRoomService.java:
```diff
@@ -25,7 +26,7 @@ public StudyRoomService(StudyRoomRepository studyRoomRepository, ParticipantRepo
     public StudyRoom createRoom(String roomName, int total, int memberSequenceId) {
         StudyRoom studyRoom = studyRoomRepository.save(roomName, total, memberSequenceId);
-        participantRepository.save(studyRoom.getRoomId(), memberSequenceId);
+        participantRepository.save(studyRoom.getRoomId(), memberSequenceId, LocalDateTime.now());
```
이처럼 현재시간의 값을 사용하는 경우 테스트가 어려워지는데요..!

현재시간 뿐 아니라 Random과 같은 제어할 수 없는 코드는 어떻게 테스트 가능하게 만들 수 있는지 참고해서 고민해보세요!
- https://www.baeldung.com/java-override-system-time
- https://jojoldu.tistory.com/676

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/room/service/StudyRoomExitService.java:
```diff
@@ -1,61 +1,62 @@
+    public StudyRoom exitRoom(int roomId, int memberSequenceId) {
+        StudyRoom studyRoom = studyRoomRepository.findByRoomId(roomId).orElseThrow(() ->
+                new RoomNotFoundException("방이 존재하지 않습니다."));
+
+        studyRoom.exitRoom();
+
+        // 방장 권한을 가진 유저일 경우 방장 권한 위임
+        if (studyRoom.isRoomManager(memberSequenceId)) {
+            transferRoomAdminRole(memberSequenceId, studyRoom);
+        }
+
+        participantRepository.delete(studyRoom.getRoomId(), memberSequenceId);
+        studyRoomRepository.update(studyRoom.getRoomId(),
+                studyRoom.getRoomName(),
+                studyRoom.getMaxParticipants(),
+                studyRoom.getCurrentParticipants(),
+                studyRoom.getManagerSequenceId());
+
+        return studyRoom;
+    }
+
+
+    // 방장 권한을 가졌을 시 권한은 다른 사용자에게 위임 하는 메서드
+    private void transferRoomAdminRole(int currentAdminSequenceID, StudyRoom studyRoom) {
+        Optional<Participant> nextAdmin =
+                findNextAdmin(currentAdminSequenceID, studyRoom.getRoomId());
+
+        nextAdmin.ifPresent(participant -> studyRoom.changeRoomManager(participant.getSeqId()));
+    }
+
+    // 권한 위임할 유저 찾는 메서드
+    private Optional<Participant> findNextAdmin(int currentAdminSequenceID, int roomId) {
+        List<Participant> participants = participantRepository.findByRoomId(roomId);
+
+        // 입장 시간이 가장 빠른 유저에게 위임
+        return participants.stream()
+                .filter(participant -> participant.getSeqId() != currentAdminSequenceID)
+                .min(Comparator.comparing(Participant::getEntryTime));
+    }
+}
```
보다 도메인 비즈니스로직이 하나의 유스케이스 단위로 응집되어 유지보수하기 좋겠네요 👍

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/room/service/StudyRoomExitService.java:
```diff
@@ -1,61 +1,62 @@
+        List<Participant> participants = participantRepository.findByRoomId(roomId);
+
+        // 입장 시간이 가장 빠른 유저에게 위임
+        return participants.stream()
+                .filter(participant -> participant.getSeqId() != currentAdminSequenceID)
+                .min(Comparator.comparing(Participant::getEntryTime));
```
이것은 사소하지만, 일급컬렉을 사용하여 기능(`longestParticipant`)을 응집했어도 좋았겠네요! (일단 현재로써는 일급컬렉션이 오버엔지니어링일수도 있겠습니다. 참고만해주세요!)

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/room/repository/ParticipantRepositoryImpl.java:
```diff
@@ -19,13 +23,14 @@ public ParticipantRepositoryImpl(JdbcTemplate jdbcTemplate) {
 
     @Override
-    public void save(int roomId, int seqId) {
+    public void save(int roomId, int seqId, LocalDateTime now) {
```
now라는 명칭보다는 entryTime이라는 이름은 어떨까요? now라는 표현은 현재시간값을 넘기는 클라이언트 입장에서 어울릴것 같아요.

wooni97 replied:
앗 그렇네요! 수정하겠습니다.

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/room/entity/StudyRoom.java:
```diff
@@ -32,6 +32,14 @@ public boolean isRoomFull() {
         return this.maxParticipants == this.currentParticipants;
     }
 
+    public void changeRoomManager(int newManagerSequenceID){
+        this.managerSequenceId = newManagerSequenceID;
+    }
+
+    public boolean isRoomManager(int memberSequenceID) {
+        return this.managerSequenceId == memberSequenceID;
+    }
+
```
StudyRoom 엔티티가 전체적으로 Participant 엔티티와 협력관계를 맺을일이 많아보여요. DB에서 객체로 로드해올 때 컬렉션 혹은 일급컬렉션으로 가져와서 StudyRoom 객체에 연관맺어주면 조금 더 객체스럽게 도메인 비즈니스 로직을 수행할 수 있겠네요! (반영할 필요는 없습니다. 참고만 해주세요)

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/room/controller/StompStudyRoomController.java:
```diff
@@ -0,0 +1,70 @@
+        messagingTemplate.convertAndSend("/subscribe/message/room/" + message.getStudyRoomID(), message);
+
+        sendRoomInfo(message.getStudyRoomID());
+        updateParticipants(message.getStudyRoomID());
```
두 기능을 응집해줘도 좋겠네요!

f-lab-inno commented on src/main/java/dev/flab/studytogether/domain/member/controller/MemberAPIController.java:
```diff
@@ -40,11 +43,20 @@ public MemberResponse join(MemberCreateRequestDto requestDto) {
+
+        Cookie sequenceIDCookie = new Cookie("sequenceID", Integer.toString(member.getSequenceId()));
+        sequenceIDCookie.setPath("/");
+        response.addCookie(sequenceIDCookie);
```
요것도 AOP혹은 인터셉터를 활용하는 방법도 있겠습니다만 조금 오버일수 있으니 하나의 내부 기능으로 내재화 해도 되겠네요!

wooni97 replied:
일단은 내부 기능으로 내재화 했습니다. AOP, 인터셉트로 적용하는 방안도 고려해볼게요!

f-lab-inno commented on src/main/java/dev/flab/studytogether/config/StompWebSocketConfig.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.web.socket.config.annotation.StompEndpointRegistry;
+import org.springframework.web.socket.config.annotation.WebSocketMessageBrokerConfigurer;
+
+@Configuration
+@EnableWebSocketMessageBroker
+public class StompWebSocketConfig implements WebSocketMessageBrokerConfigurer {
+
+    @Override
+    public void registerStompEndpoints(StompEndpointRegistry registry) {
+
+        // "stomp-room" 는 웹소캣 또는 SocketJS Client가 웹소캣 핸드셰이크 커넥션을 생성할 경로이다.
+        registry.addEndpoint("/stomp-room")
+                .setAllowedOrigins("http://localhost:8080")
+                .withSockJS();
+    }
+
+    @Override
+    public void configureMessageBroker(MessageBrokerRegistry config) {
+           config.enableSimpleBroker("/subscribe");
+           config.setApplicationDestinationPrefixes("/publish");
+    }
+}
```
STOMP 로 웹소켓 서버를 잘 구현해주셨네요.

STOMP가 웹소켓 구현을 어떻게 추상화하여 제공하는지 Raw한 WebSocket에 비해 어떤 장점이 있는지 알아보세요!
- https://kellis.tistory.com/65
- https://www.geeksforgeeks.org/stomp-protocol/
- https://www.rfwireless-world.com/Terminology/STOMP-architecture.html

f-lab-inno commented on build.gradle:
```diff
@@ -41,6 +41,15 @@ dependencies {
 
+    implementation 'org.springframework.boot:spring-boot-starter-websocket'
+
+    // jQuery WebJar
+    implementation 'org.webjars:jquery:3.6.0'
+
+    // SockJS WebJar
+    implementation 'org.webjars:sockjs-client:1.0.2'
```
오? 클라이언트 의존성처럼 보이는데, gradle에 걸어놓으신 이유가 있을까요?

wooni97 replied:
프로젝트 내의 모든 의존성들을 한 곳에서 일관되게 관리하는게 편리하다고 생각해 gradle을 사용했습니다..!

f-lab-inno replied:
그런 관점이 있군요, 좋습니다!

wooni97 committed:
refactor: setCookies 메서드 추가

- 기존 login 메서드에서 쿠키 생성 로직을 별도의 메서드 기능으로 분리합니다.

wooni97 committed:
refactor: var를 let으로 변경

- ES6 표준 문법을 따라 var 선언은 let으로 변경합니다.

wooni97 committed:
refactor: exit 메서드 삭제

- Study Room 퇴장 로직을 StudyRoomExitService로 분리함에 따라 exit 메서드를 삭제합니다.

wooni97 committed:
refactor: 변수명 변경

- entryTime이 입장 시간의 의미를 더 명확히 나타내기에 변경합니다.

wooni97 committed:
refactor: Study Room 정보 및 참여자 업데이트 메서드 통합

- 'sendRoomInfo'와 'updateParticipants' 메서드를 하나로 통합하여 방 업데이트를 처리하는 'updateRoomAndParticipants' 메서드를 추가합니다.
